neo4j_python_migrations/executor.py

```diff
@@ -66,16 +66,20 @@ def migrate(self, on_apply: Optional[Callable[[Migration], None]] = None) -> Non
 
         for migration in analyzing_result.pending_migrations:
             with self.driver.session(database=self.database) as session:
-                start_time = time.monotonic()
-                migration.apply(session)
-                duration = time.monotonic() - start_time
-            self.dao.add_migration(
-                migration,
-                duration,
-            )
+                with session.begin_transaction() as tx:
+
+                    start_time = time.monotonic()
+                    migration.apply(tx)
+                    duration = time.monotonic() - start_time
+
+                    self.dao.add_migration(
+                        migration,
+                        duration,
+                        tx
+                    )
```

Comment:
It would be cool to do the migration and update the migration graph in a single transaction. But earlier and now I got a problem if a schema was performed during migration.

Here is an example:
```
from neo4j import GraphDatabase

with (
    GraphDatabase.driver(
        "neo4j://localhost:7687",
        auth=("neo4j", "test1234"),
    ) as driver,
    driver.session() as session,
    session.begin_transaction() as tx,
):
    # Can be considered as a migration code
    tx.run("CREATE CONSTRAINT foobar FOR (n:User) REQUIRE n.name IS UNIQUE")

    # Can be considered as a migration graph update code
    tx.run("MERGE (:User {name: 'Root'})")  
```

`neo4j.exceptions.Forbidden: {code: Neo.ClientError.Transaction.ForbiddenDueToTransactionType} {message: Tried to execute Write query after executing Schema modification}`

Because of this problem, I had to not use the solution with a single transaction. Maybe you have some ideas for a solution?

Reply:
Ouch, that's a good point, I don't know why I didn't come across that. Still trying to get to grips with do's and don't in Neo4j, but it's clear that these two operations need to happen in different transactions.

Multiple ideas come to mind, none of them perfect:

1. Do the migration graph update in a transaction before the migration, and manually roll it back if migration failed.
2. Run two async transactions, where both are rolled back if either fails.

It might also be that ultimately, the best solution is to make the "update migration graph"-code as robust as possible. After all, the root problem of that caused me to dig into this was the fact that it failed when running against a db with authentication disabled.

Reply:
Thanks for the suggestions, I will try to implement  "dry-run" the migration graph upgrade by rolling back the transaction